src/Middleware/FailureHandling/FailureEnvelope.php

```diff
@@ -4,32 +4,28 @@
 
 namespace Yiisoft\Queue\Middleware\FailureHandling;
 
-use Yiisoft\Queue\Message\EnvelopeInterface;
-use Yiisoft\Queue\Message\EnvelopeTrait;
+use Yiisoft\Arrays\ArrayHelper;
+use Yiisoft\Queue\Message\Envelope;
 use Yiisoft\Queue\Message\MessageInterface;
 
-final class FailureEnvelope implements EnvelopeInterface
+final class FailureEnvelope extends Envelope
 {
-    use EnvelopeTrait;
-
     public const FAILURE_META_KEY = 'failure-meta';
 
     public function __construct(
-        private MessageInterface $message,
-        private array $meta = [],
+        MessageInterface $message,
+        private readonly array $failureMeta = [],
     ) {
+        parent::__construct($message);
     }
 
     public static function fromMessage(MessageInterface $message): self
     {
         return new self($message, $message->getMetadata()[self::FAILURE_META_KEY] ?? []);
     }
 
-    public function getMetadata(): array
+    protected function getEnvelopeMetadata(): array
     {
-        $meta = $this->message->getMetadata();
-        $meta[self::FAILURE_META_KEY] = array_merge($meta[self::FAILURE_META_KEY] ?? [], $this->meta);
-
-        return $meta;
+        return [self::FAILURE_META_KEY => ArrayHelper::merge($this->message->getMetadata()[self::FAILURE_META_KEY] ?? [], $this->failureMeta)];
```

Comment:
as you can see `getEnvelopeMetadata` method must provide all the collected metadata, not only the owned

Reply:
For all metadata retrieval we have `Envelope::getMetadata()`. The `getEnvelopeMetadata()` method returns envelope metadata only.